transport/tls-record-frag/writer.go

```diff
@@ -0,0 +1,132 @@
+
+func NewWriter(writer io.Writer, prefixBytes int32) *tlsRecordFragWriter {
+	return &tlsRecordFragWriter{writer, prefixBytes}
+}
+
+func (w *tlsRecordFragWriter) dontFrag(first []byte, source io.Reader) (written int64, err error) {
+	tmp, err := w.writer.Write(first)
+	written = int64(tmp)
+	w.prefixBytes = 0
+	if err != nil {
+		return written, err
+	}
+	n, err := io.Copy(w.writer, source)
+	written += n
+	return written, err
+}
+
+func (w *tlsRecordFragWriter) ReadFrom(source io.Reader) (written int64, err error) {
+	if 0 < w.prefixBytes {
+		var recordHeader [5]byte
+		_, err := io.ReadFull(source, recordHeader[:])
+		if err != nil {
+			return 0, err
+		}
+		recordLength := int32(binary.BigEndian.Uint16(recordHeader[3:]))
+		if w.prefixBytes >= recordLength {
+			return w.dontFrag(recordHeader[:], source)
+		}
+		if recordLength > maxRecordLength {
+			return w.dontFrag(recordHeader[:], source)
+		}
+		// Allocate buffer that fits the entire record after the split (2*header + payload).
+		buf := make([]byte, recordLength+10)
+		header := recordHeader[:3]
+
+		copy(buf, header)
+		binary.BigEndian.PutUint16(buf[3:], uint16(w.prefixBytes))
+		n2, err := io.ReadFull(source, buf[5:5+w.prefixBytes])
+		if err != nil {
+			w.prefixBytes = 0
+			return 0, err
+		}
+
+		copy(buf[5+n2:], header)
+		binary.BigEndian.PutUint16(buf[5+n2+3:], uint16(recordLength-w.prefixBytes))
+		_, err = io.ReadFull(source, buf[10+w.prefixBytes:])
+		if err != nil {
+			w.prefixBytes = 0
+			return 0, err
+		}
+
+		tmp, err := w.writer.Write(buf)
+		w.prefixBytes = 0
+		written = int64(tmp)
+		if err != nil {
+			return written, err
+		}
+	}
+	n, err := io.Copy(w.writer, source)
+	written += n
+	return written, err
+}
+
+func (w *tlsRecordFragWriter) Write(data []byte) (written int, err error) {
+	if 0 < w.prefixBytes {
+		length := int32(len(data))
+		if w.prefixBytes+5 >= length {
+			w.prefixBytes = 0
+			return w.writer.Write(data)
+		}
+
+		recordLength := int32(binary.BigEndian.Uint16(data[3:]))
+		remainder := data[5:]
+		remainderLength := length - 5
+		hasPartialRecord := recordLength > remainderLength
+		hasMultipleRecords := recordLength < remainderLength
+		isRecordOverflow := recordLength > maxRecordLength
+
+		if hasPartialRecord || w.prefixBytes == recordLength || isRecordOverflow {
+			w.prefixBytes = 0
+			return w.writer.Write(data)
+		}
+		//Need fragmentation, allocate data + header*1
+		buf := make([]byte, length+5)
+		header := data[:3]
+		record1 := remainder[:w.prefixBytes]
+		record2 := remainder[w.prefixBytes:recordLength]
+
+		copy(buf, header)
+		binary.BigEndian.PutUint16(buf[3:], uint16(w.prefixBytes))
+		copy(buf[5:], record1)
+
+		copy(buf[5+w.prefixBytes:], header)
+		binary.BigEndian.PutUint16(buf[5+3+w.prefixBytes:], uint16(len(record2)))
+		copy(buf[5+5+w.prefixBytes:], record2)
+
+		if hasMultipleRecords {
+			copy(buf[5+5+recordLength:], remainder[recordLength:])
+		}
+
+		w.prefixBytes = 0
+		return w.writer.Write(buf)
```

Comment:
We need to discount the bytes added here too, so we output the number of input bytes consumed.